rfcs/20210415-gpu-only/README.md

```diff
@@ -0,0 +1,185 @@
+oneDNN only for a CPU runtime the GPU runtime always requires a CPU runtime to be
+specified. For some customers the size of the oneDNN library matters a lot therefore
+if the customers only need a GPU runtime then having a redundant CPU runtime
+becomes a problem. One of the customers is OpenVINO whose request is the main motivation
+for introducing an ability to build oneDNN only for GPU runtime.
+
+## Proposal
+The proposal is to introduce the following GPU only configurations:
+* gpu_dpcpp (`DNNL_CPU_RUNTIME=NONE DNNL_GPU_RUNTIME=DPCPP`)
+* gpu_ocl (`DNNL_CPU_RUNTIME=NONE DNNL_GPU_RUNTIME=OCL`)
+
+We need to support them both because the `gpu_dpcpp` configuration is expected to
+be the main one for GPU but currently OpenVINO is still using OpenCL runtime.
+Fortunately, the vast majority of the code is shared between OCL and DPCPP runtimes
+therefore enabling both should not increase amount of work significantly.
+
+Currently, there is no plan to include these configurations in oneAPI as well as
+enable and test GPU only configuration for NVIDIA.
+
+
+### API
+
+oneDNN provides some APIs which are CPU specific or implemented only for CPU.
+There are two options to deal with it:
+
+#### Option 1: Removing CPU Specific API
+* Inside the library the API will be guarded by macros (e.g `DNNL_GPU_ONLY`)
+* The `dnnl.h` and `dnnl.hpp` headers can be configured by CMake in the following way:
+    ```cpp
+    // dnnl.h
+
+    #define DNNL_GPU_ONLY 1 // this macro is substituted by CMake during configuration
+    // ... //
+    #if DNNL_GPU_ONLY
+    dnnl_status_t DNNL_API dnnl_set_max_cpu_isa(dnnl_cpu_isa_t isa);
+    #endif
+    // ... //
+    #undef DNNL_GPU_ONLY // end of dnnl.h
+    ```
+* Basically, this means that a user will not be able to call any CPU specific API
+
+#### Option 2: API Stays Intact
+* Inside the library the API will be adjusted to return a status in the case of GPU
+only configuration
+* The status can be consistent for all CPU specific API or can be different
+depending on the API
+* For example, for `dnnl_set_max_cpu_isa` oneDNN
+can return `runtime_error` because it's a dedicated API for CPU but for
+`dnnl_gemm_s8s8s32` oneDNN can return `unimplemented`.
+* The GTest will be adjusted to make sure that such APIs return proper status
+
+The suggestion is to go with the second option as this is better aligned with the
+runtime agnostic approach that oneDNN uses for almost all APIs. Also, it allows
+applications that use CPU specific API under a runtime condition to link with
+oneDNN.
+
+### Cross-Engine Reorder
+
+Since CPU engine is not available in GPU only configuration the cross-engine
+reorder capability will be limited to GPU -> GPU case. The only exception is when
+a user creates `dnnl::memory` using a pointer (handle) to shared USM. In this case,
+the behavior will be similar to CPU -> GPU and GPU -> CPU cases.
+
+OpenVINO mentioned that they used the cross-engine reorder to avoid dealing with
```

Comment:
Actually we don't use cross engine reorder in cldnn plugin. So maybe the proper solution would be disabling cross engine reorder if oneDNN is built with single engine support?

Reply:
Thanks for clarifying this. Technically, the cross-engine reorder can be disabled for GPU only configurations, but I think we should still provide a way to reorder from one GPU device to another. Implementation of a cross-engine reorder is expected to be lightweight (from binary size perspective) because it consists of a regular reorder and a call to the runtime specific API to copy data.